src/osx/cocoa/textctrl.mm

```diff
@@ -1142,80 +1367,256 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
+                        [paraStyle setMaximumLineHeight:0];
+                        break;
+
+                    case wxTEXT_ATTR_LINE_SPACING_MODE_EXACT:
+                        [paraStyle setLineHeightMultiple:1];
+                        [paraStyle setMinimumLineHeight:PointsFromInternalUnits(style.GetLineSpacing(), style.GetUnitsPerPoint())];
+                        [paraStyle setMaximumLineHeight:[paraStyle minimumLineHeight]];
+                        break;
+                }
+            }
+
+            if ( style.HasTextEffect(wxTEXT_ATTR_EFFECT_RTL) )
+            {
+                if ( textsEffects & wxTEXT_ATTR_EFFECT_RTL )
+                    [paraStyle setBaseWritingDirection:NSWritingDirectionRightToLeft];
+                else
+                    [paraStyle setBaseWritingDirection:NSWritingDirectionLeftToRight];
+            }
+
+            if ( style.HasTextEffect(wxTEXT_ATTR_EFFECT_SUPPRESS_HYPHENATION) )
+            {
+                if ( textsEffects & wxTEXT_ATTR_EFFECT_SUPPRESS_HYPHENATION )
+                    [paraStyle setHyphenationFactor:0.0];
+                else
+                    [paraStyle setHyphenationFactor:1.0];
+            }
+
+            if ( style.HasTabs() )
+            {
+                while ([[paraStyle tabStops] count] > 0)
+                {
+                    NSTextTab* tab = [[paraStyle tabStops] objectAtIndex:0];
+                    [paraStyle removeTabStop:tab];
+                }
+
+                wxArrayInt tabs = style.GetTabs();
+                for (int i = 0; i < tabs.size(); i++)
+                {
+                    long position = tabs[i];
+                    NSTextTab* tab = [[NSTextTab alloc]
+                                      initWithType:NSLeftTabStopType
+                                      location:PointsFromInternalUnits(position, style.GetUnitsPerPoint())];
+
+                    [paraStyle addTabStop:tab];
+                }
             }
-            [dict setObject:[NSNumber numberWithInt:( underlineStyle )] forKey:NSUnderlineStyleAttributeName];
-            wxColour colour = style.GetUnderlineColour();
-            if ( colour.IsOk() )
+
+            [storage removeAttribute:NSParagraphStyleAttributeName range:range];
+            [storage addAttribute:NSParagraphStyleAttributeName value:paraStyle range:range];
+            [paraStyle release];
+
+            if ( style.HasParagraphUserData() )
             {
-                [dict setValue:colour.OSXGetNSColor() forKey:NSUnderlineColorAttributeName];
+                [storage removeAttribute:wxNSParagraphUserDataAttributeName range:range];
+                [storage addAttribute:wxNSParagraphUserDataAttributeName value:[NSNumber numberWithLong:style.GetParagraphUserData()] range:range];
             }
-            [storage addAttributes:dict range:range];
-            [dict release];
         }
     }
+    @catch (NSException *exception)
```

Comment:
Why do we do need to do this? I.e. why should any exceptions happen here and, if they do, why do we just log them without doing anything else?

Reply:
There are subtle cases where the selection range is not valid, or where the font selected can't be applied to the text because it doesn't support the selected characters.  Letting those exceptions out of the Objective-C code is bad.